pkg/handler/collector/github/github.go

```diff
@@ -0,0 +1,327 @@
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/logging"
+)
+
+const (
+	GithubCollector = "GithubCollector"
+)
+
+func defaultAssetSuffixes() []string {
+	return []string{".jsonl"}
+}
+
+type TagOrLatest interface {
+	IsTagOrLatest()
+}
+
+type Tag struct {
+	Value string
+}
+
+type Latest struct{}
+
+func (t *Tag) IsTagOrLatest()    {}
+func (l *Latest) IsTagOrLatest() {}
+
+// Compile time checks
+var _ TagOrLatest = &Tag{}
+var _ TagOrLatest = &Latest{}
+
+type githubCollector struct {
+	poll              bool
+	interval          time.Duration
+	client            githubclient.GithubClient
+	repoToReleaseTags map[client.Repo][]TagOrLatest
+	assetSuffixes     []string
+	collectDataSource datasource.CollectSource
+}
+
+type Config struct {
+	Poll              bool
+	Interval          time.Duration
+	Client            githubclient.GithubClient
+	RepoToReleaseTags map[client.Repo][]TagOrLatest
+	AssetSuffixes     []string
+	CollectDataSource datasource.CollectSource
+}
+
+type Opt func(*githubCollector)
+
+func NewGithubCollector(opts ...Opt) (*githubCollector, error) {
+	g := &githubCollector{
+		poll:              false,
+		interval:          0,
+		client:            nil,
+		repoToReleaseTags: map[client.Repo][]TagOrLatest{},
+		assetSuffixes:     defaultAssetSuffixes(),
+		collectDataSource: nil,
+	}
+
+	for _, opt := range opts {
+		opt(g)
+	}
+
+	if g.client == nil {
+		return nil, fmt.Errorf("no github client provided for collector")
+	}
+	if len(g.assetSuffixes) == 0 {
+		return nil, fmt.Errorf("no asset suffixes for github collector")
+	}
+	if len(g.repoToReleaseTags) == 0 && g.collectDataSource == nil {
+		return nil, fmt.Errorf("no repos and releases to collect nor any data source for future subscriptions")
+	}
+	return g, nil
+}
+
+func WithPolling(interval time.Duration) Opt {
+	return func(g *githubCollector) {
+		g.poll = true
+		g.interval = interval
+	}
+}
+
+func WithClient(client githubclient.GithubClient) Opt {
+	return func(g *githubCollector) {
+		g.client = client
+	}
+}
+
+func WithRepoToReleaseTags(repoToReleaseTags map[client.Repo][]TagOrLatest) Opt {
+	return func(g *githubCollector) {
+		g.repoToReleaseTags = repoToReleaseTags
+	}
+}
+
+func WithAssetSuffixes(assetSuffixes []string) Opt {
+	return func(g *githubCollector) {
+		g.assetSuffixes = assetSuffixes
+	}
+}
+
+func WithCollectDataSource(collectDataSource datasource.CollectSource) Opt {
+	return func(g *githubCollector) {
+		g.collectDataSource = collectDataSource
+	}
+}
+
+// RetrieveArtifacts get the artifacts from the collector source based on polling or one time
+func (g *githubCollector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
+	err := g.populateRepoToReleaseTags(ctx)
+	if err != nil {
+		return err
+	}
+	if g.poll {
+		for repo, tags := range g.repoToReleaseTags {
+			g.fetchAssets(ctx, repo.Owner, repo.Repo, tags, docChannel)
+			time.Sleep(g.interval)
```

Comment:
should the poll be between fetches of releases? if not the poll time scales with the number of tags?